### PR TITLE
add mds metric

### PR DESCRIPTION
**Pull request by cw123**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by cw123** [comment]
recheck

File: curvefs/src/mds/topology/topology.cpp [hunk 1]
```diff
@@ -749,6 +761,30 @@ std::list<ZoneIdType> TopologyImpl::GetZoneInPool(PoolIdType id,
     return ret;
 }
 
+std::vector<CopySetIdType> TopologyImpl::GetCopySetsInPool(
+    PoolIdType poolId, CopySetFilter filter) const {
+    std::vector<CopySetIdType> ret;
+    ReadLockGuard rlockCopySet(copySetMutex_);
+    for (auto it : copySetMap_) {
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology.cpp:768]: done

File: curvefs/src/mds/topology/topology.cpp [hunk 2]
```diff
@@ -749,6 +761,30 @@ std::list<ZoneIdType> TopologyImpl::GetZoneInPool(PoolIdType id,
...
+        }
+    }
+    return ret;
+}
+
+std::vector<CopySetInfo> TopologyImpl::GetCopySetInfosInPool(
+    PoolIdType poolId, CopySetFilter filter) const {
+    std::vector<CopySetInfo> ret;
+    ReadLockGuard rlockCopySet(copySetMutex_);
+    for (auto it : copySetMap_) {
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology.cpp:780]: done

File: curvefs/src/mds/topology/topology_config.h [hunk 3]
```diff
@@ -42,14 +42,17 @@ struct TopologyOption {
...
+    uint32_t UpdateMetricIntervalSec;
 
     TopologyOption()
         : topologyUpdateToRepoSec(0),
           partitionNumberInCopyset(256),
           idNumberInPartition(16777216),
           choosePoolPolicy(0),
           createCopysetNumber(10),
-          createPartitionNumber(1) {}
+          createPartitionNumber(1),
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_config.h:54]: done

File: curvefs/src/mds/topology/topology_metric.cpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
+/*
+ *  Copyright (c) 2020 NetEase Inc.
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.cpp:2]: done

File: curvefs/src/mds/topology/topology_metric.cpp [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curvefs
+ * Created Date: Thu Jun 27 2019
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.cpp:19]: done

File: curvefs/src/mds/topology/topology_metric.cpp [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+ * Author: xuchaojie
+ */
+
+#include "curvefs/src/mds/topology/topology_metric.h"
+
+#include <memory>
+#include <set>
+#include <list>
+
+// #include "curvefs/src/mds/copyset/copyset_validation.h"
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.cpp:29]: done

File: curvefs/src/mds/topology/topology_metric.h [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curvefs
+ * Created Date: Thu Jun 27 2019
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.h:19]: done

File: curvefs/src/mds/topology/topology_metric.h [hunk 8]
```diff
@@ -0,0 +1,177 @@
+/*
+ *  Copyright (c) 2020 NetEase Inc.
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.h:2]: done

File: curvefs/src/mds/topology/topology_metric.h [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+using ::curve::common::InterruptibleSleeper;
+
+namespace curvefs {
+namespace mds {
+namespace topology {
+
+struct MetaServerMetric {
+    const std::string kTopologyMetaServerMetricPrefix =
+        "topology_metric_metaserver_Id_";
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.h:46]: done

File: curvefs/src/mds/topology/topology_metric.h [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+          memoryUsed(kTopologyMetaServerMetricPrefix,
+                     std::to_string(msId) + "_memoryUsed", 0),
+          partitionNum(kTopologyMetaServerMetricPrefix,
+                       std::to_string(msId) + "_partitionNum", 0) {}
+};
+
+using MetaServerMetricPtr = std::unique_ptr<MetaServerMetric>;
+
+struct PoolMetric {
+    const std::string kTopologyPoolMetricPrefix = "topology_metric_Pool_";
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.h:82]: done

File: curvefs/src/mds/topology/topology_metric.h [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                     std::to_string(msId) + "_copysetnum", 0),
+          leaderNum(kTopologyMetaServerMetricPrefix,
+                    std::to_string(msId) + "_leadernum", 0),
+          diskCapacity(kTopologyMetaServerMetricPrefix,
+                       std::to_string(msId) + "_diskCapacity", 0),
+          diskUsed(kTopologyMetaServerMetricPrefix,
+                   std::to_string(msId) + "_diskUsed", 0),
+          memoryUsed(kTopologyMetaServerMetricPrefix,
+                     std::to_string(msId) + "_memoryUsed", 0),
+          partitionNum(kTopologyMetaServerMetricPrefix,
```
[the remark this answers is not in the document]
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.h:75]: done

File: curvefs/src/mds/topology/topology_metric.cpp [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+        it->second->inodeNum.set_value(totalInodeNum);
+        it->second->dentryNum.set_value(totalDentryNum);
+        it->second->partitionNum.set_value(partitions.size());
+
+
+        uint64_t totalDiskCapacity = 0;
+        uint64_t totalDiskUsed = 0;
+
+        // process the metric of metaserver.
+        for (auto cm : metaServerMetricInfo) {
```
Comment (SeanHai) [inline curvefs/src/mds/topology/topology_metric.cpp:98]: for (const auto &cm : metaServerMetricInfo)

File: curvefs/src/mds/topology/topology_metric.cpp [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    return;
+}
+
+void TopologyMetricService::CalcMetaServerMetrics(
+    const std::vector<CopySetInfo> &copysets,
+    std::map<MetaServerIdType, MetaServerMetricInfo> *msMetricInfoMap) {
+    for (auto &cs : copysets) {
+        for (auto &msId : cs.GetCopySetMembers()) {
```
Comment (SeanHai) [inline curvefs/src/mds/topology/topology_metric.cpp:145]: for (const auto &cs : copysets)
    for (const auto &msId : cs.GetCopySetMembers())

Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.cpp:145]: done

File: curvefs/src/mds/topology/topology_metric.cpp (hunk 12 above)
Comment (cw123) [reply curvefs/src/mds/topology/topology_metric.cpp:98]: done

File: curvefs/src/mds/topology/topology_metric.cpp [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+        }
+        it->second->inodeNum.set_value(totalInodeNum);
+        it->second->dentryNum.set_value(totalDentryNum);
+        it->second->partitionNum.set_value(partitions.size());
+
+        uint64_t totalDiskCapacity = 0;
+        uint64_t totalDiskUsed = 0;
+
+        // process the metric of metaserver.
+        for (const auto cm : metaServerMetricInfo) {
```
Comment (xu-chaojie) [inline curvefs/src/mds/topology/topology_metric.cpp:97]: const auto cm --> const auto &cm

**Review by xu-chaojie — APPROVED** [review]

**Review by SeanHai — APPROVED** [review]

**Comment by cw123** [comment]
recheck

**Pull request merged**